File: src/main/java/destiny/astrology/classical/rules/accidentalDignities/JoyHouse.java
/**
 * @author smallufo
 * Created on 2007/12/30 at 上午 12:07:11
 */
package destiny.astrology.classical.rules.accidentalDignities;

import destiny.astrology.Horoscope;
import destiny.astrology.Planet;
import org.jetbrains.annotations.NotNull;
import org.jooq.lambda.tuple.Tuple;
import org.jooq.lambda.tuple.Tuple2;

import java.util.Optional;

/**
 * 喜樂宮 Joy House. 
 * Mercory in 1st. 
 * Moon in 3rd. 
 * Venus in 5th. 
 * Mars in 6th. 
 * Sun in 9th. 
 * Jupiter in 11th. 
 * Saturn in 12th.
 */
public final class JoyHouse extends Rule {

  public JoyHouse()
  {
  }

  @Override
  protected Optional<Tuple2<String, Object[]>> getResult(Planet planet, @NotNull Horoscope h) {

    return h.getHouse(planet)
      .filter(house ->
          (planet == Planet.MERCURY && house == 1) ||
          (planet == Planet.MOON && house == 3) ||
          (planet == Planet.VENUS && house == 5) ||
          (planet == Planet.MARS && house == 6) ||
          (planet == Planet.SUN && house == 9) ||
          (planet == Planet.JUPITER && house == 11) ||
          (planet == Planet.SATURN && house == 12)
      )
      .map(house -> Tuple.tuple("comment", new Object[]{planet, house}));

  }

}

File: src/main/java/destiny/astrology/classical/rules/accidentalDignities/JoyHouse.kt
/**
 * @author smallufo
 * Created on 2007/12/30 at 上午 12:07:11
 */
package destiny.astrology.classical.rules.accidentalDignities

import destiny.astrology.IHoroscopeModel
import destiny.astrology.Planet
import destiny.astrology.Planet.*

/**
 * 喜樂宮 Joy House.
 * Mercury in 1st.
 * Moon in 3rd.
 * Venus in 5th.
 * Mars in 6th.
 * Sun in 9th.
 * Jupiter in 11th.
 * Saturn in 12th.
 */
class JoyHouse : AccidentalRule() {

  override fun getResult(planet: Planet, h: IHoroscopeModel): Pair<String, Array<Any>>? {
    return h.getHouse(planet)
      ?.takeIf { house ->
        planet === MERCURY && house == 1 ||
          planet === MOON && house == 3 ||
          planet === VENUS && house == 5 ||
          planet === MARS && house == 6 ||
          planet === SUN && house == 9 ||
          planet === JUPITER && house == 11 ||
          planet === SATURN && house == 12
      }?.let { house -> "comment" to arrayOf(planet, house) }
  }
}
